File: src/main/java/me/geek/tom/serverutils/commands/BroadcastCommand.java
package me.geek.tom.serverutils.commands;

import com.mojang.brigadier.CommandDispatcher;
import com.mojang.brigadier.context.CommandContext;
import me.geek.tom.serverutils.TomsServerUtils;
import net.minecraft.network.MessageType;
import net.minecraft.server.MinecraftServer;
import net.minecraft.server.command.ServerCommandSource;
import net.minecraft.text.LiteralText;
import net.minecraft.text.Text;
import net.minecraft.util.Util;

import static com.mojang.brigadier.arguments.StringArgumentType.getString;
import static com.mojang.brigadier.arguments.StringArgumentType.greedyString;
import static net.minecraft.command.argument.TextArgumentType.getTextArgument;
import static net.minecraft.command.argument.TextArgumentType.text;
import static net.minecraft.server.command.CommandManager.argument;
import static net.minecraft.server.command.CommandManager.literal;

public class BroadcastCommand {
    public static void register(CommandDispatcher<ServerCommandSource> dispatcher) {
        dispatcher.register(literal("broadcast")
                .requires(s -> s.hasPermissionLevel(2))
                .then(literal("raw")
                        .then(argument("message", text())
                                .executes(ctx -> {
                                    Text text = getTextArgument(ctx, "message");
                                    broadcastMessage(ctx, text);
                                    return 0;
                                })
                        )
                ).then(argument("message", greedyString())
                        .executes(ctx -> {
                            String message = getString(ctx, "message");
                            Text text = new LiteralText(message);
                            broadcastMessage(ctx, text);
                            return 0;
                        })
                )
        );
    }

    private static void broadcastMessage(CommandContext<ServerCommandSource> ctx, Text text) {
        MinecraftServer server = ctx.getSource().getMinecraftServer();
        server.getPlayerManager().broadcastChatMessage(text, MessageType.CHAT, Util.NIL_UUID);
        TomsServerUtils.broadcast(text);
    }
}

File: src/main/kotlin/commands/BroadcastCommand.kt
package me.geek.tom.serverutils.commands

import com.github.p03w.aegis.aegisCommand
import com.mojang.brigadier.CommandDispatcher
import com.mojang.brigadier.arguments.StringArgumentType.getString
import com.mojang.brigadier.context.CommandContext
import me.geek.tom.serverutils.broadcast
import net.minecraft.command.argument.TextArgumentType.getTextArgument
import net.minecraft.network.MessageType
import net.minecraft.server.command.ServerCommandSource
import net.minecraft.text.LiteralText
import net.minecraft.text.Text
import net.minecraft.util.Util

fun registerBroadcastCommand(dispatcher: CommandDispatcher<ServerCommandSource>) {
    dispatcher.register(aegisCommand("broadcast") {
        literal("raw") {
            text("message") {
                executes { ctx ->
                    val text = getTextArgument(ctx, "message")
                    broadcastMessage(ctx, text)
                }
            }
        }

        greedyString("message") {
            executes { ctx ->
                val message = getString(ctx, "message")
                val text = LiteralText(message)
                broadcastMessage(ctx, text)
            }
        }
    })
}

private fun broadcastMessage(ctx: CommandContext<ServerCommandSource>, text: Text) {
    val server = ctx.source.server
    server.playerManager.broadcastChatMessage(text, MessageType.CHAT, Util.NIL_UUID)
    broadcast(text)
}
